pkg/ssh/connection_native.go

```diff
@@ -0,0 +1,182 @@
+package ssh
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"regexp"
+	"strings"
+
+	"github.com/containers/common/pkg/config"
+)
+
+func nativeConnectionCreate(options ConnectionCreateOptions) error {
+	var match bool
+	var err error
+	if match, err = regexp.Match("^[A-Za-z][A-Za-z0-9+.-]*://", []byte(options.Path)); err != nil {
+		return fmt.Errorf("invalid destination: %w", err)
+	}
+
+	if !match {
+		options.Path = "ssh://" + options.Path
+	}
+
+	if len(options.Socket) > 0 {
+		options.Path += options.Socket
+	}
+
+	dst, uri, err := Validate(options.User, options.Path, options.Port, options.Identity)
+	if err != nil {
+		return err
+	}
+
+	// test connection
+	ssh, err := exec.LookPath("ssh")
+	if err != nil {
+		return fmt.Errorf("no ssh binary found")
+	}
+
+	if strings.Contains(uri.Host, "/run") {
+		uri.Host = strings.Split(uri.Host, "/run")[0]
+	}
+	conf, err := config.Default()
+	if err != nil {
+		return err
+	}
+
+	args := []string{uri.User.String() + "@" + uri.Hostname()}
+
+	if len(dst.Identity) > 0 {
+		args = append(args, "-i", dst.Identity)
+	}
+	if len(conf.Engine.SSHConfig) > 0 {
+		args = append(args, "-F", conf.Engine.SSHConfig)
+	}
+
+	output := &bytes.Buffer{}
+	args = append(args, "podman", "info", "--format", "json")
+	info := exec.Command(ssh, args...)
+	info.Stdout = output
+	err = info.Run()
+	if err != nil {
+		return err
+	}
+
+	remoteInfo := &Info{}
+	if err := json.Unmarshal(output.Bytes(), &remoteInfo); err != nil {
+		return fmt.Errorf("failed to parse 'podman info' results: %w", err)
+	}
+
+	if remoteInfo.Host.RemoteSocket == nil || len(remoteInfo.Host.RemoteSocket.Path) == 0 {
+		return fmt.Errorf("remote podman %q failed to report its UDS socket", uri.Host)
+	}
+
+	cfg, err := config.ReadCustomConfig()
+	if err != nil {
+		return err
+	}
+	if options.Default {
+		cfg.Engine.ActiveService = options.Name
+	}
```

Comment:
This part does not exist in `golangConnectionCreate`. Isn’t that a bug?

(Really it seems to me that the level of abstraction of this function is incorrect — a lot of the code should not be duplicated between native/golang here, _and_ should not be duplicated with Podman’s `connection/add`.)